File: 13-02-2026/problem_6_coin_change/correct/coin_change.c
// PROBLEM 3: COIN CHANGE - CORRECT C
// ===================================================
#include <stdio.h>
#define INF 1000000000

int min(int a, int b) {
    return a < b ? a : b;
}

int main() {
    int n, amt;
    scanf("%d %d", &n, &amt);
    int c[55];
    
    for (int i = 0; i < n; i++) {
        scanf("%d", &c[i]);
    }
    
    static int dp[100005];
    
    for (int i = 0; i <= amt; i++) {
        dp[i] = INF;
    }
    
    dp[0] = 0;
    
    for (int i = 0; i < n; i++) {
        for (int a = c[i]; a <= amt; a++) {
            if (dp[a - c[i]] != INF) {
                dp[a] = min(dp[a], dp[a - c[i]] + 1);
            }
        }
    }
    
    printf("%d\n", dp[amt] >= INF ? -1 : dp[amt]);
    return 0;
}


// ===================================================

File: 13-02-2026/problem_6_coin_change/correct/coin_change.java
// PROBLEM 3: COIN CHANGE - CORRECT JAVA
// ===================================================
import java.util.*;

class Solution {
    static final int INF = 1000000000;
    
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int amt = sc.nextInt();
        
        int[] coins = new int[n];
        for (int i = 0; i < n; i++) {
            coins[i] = sc.nextInt();
        }
        
        int[] dp = new int[amt + 1];
        Arrays.fill(dp, INF);
        dp[0] = 0;
        
        for (int i = 0; i < n; i++) {
            for (int a = coins[i]; a <= amt; a++) {
                if (dp[a - coins[i]] != INF) {
                    dp[a] = Math.min(dp[a], dp[a - coins[i]] + 1);
                }
            }
        }
        
        System.out.println(dp[amt] >= INF ? -1 : dp[amt]);
        sc.close();
    }
}


// ===================================================